File: test_drive/lib/widgets/hybrid_editor/core/transactions/update_block_content_operation.dart
import '../document.dart';
import 'operation.dart';

/// Operation to update the content of a block
class UpdateBlockContentOperation implements Operation {
  final int blockIndex;
  final String oldContent;
  final String newContent;

  const UpdateBlockContentOperation({
    required this.blockIndex,
    required this.oldContent,
    required this.newContent,
  });

  @override
  OperationType get type => OperationType.updateBlock;

  @override
  Document apply(Document document) {
    final block = document[blockIndex];
    return document.updateBlock(
      blockIndex,
      block.copyWith(content: newContent),
    );
  }

  @override
  Operation inverse(Document document) {
    return UpdateBlockContentOperation(
      blockIndex: blockIndex,
      oldContent: newContent,
      newContent: oldContent,
    );
  }
}

File: native/components/editor/core/operations/UpdateBlockContentOperation.ts
import { Document, updateBlock } from '../Document';
import { copyBlock } from '../BlockNode';
import { Operation, OperationType } from './Operation';

/// Operation to update the content of a block
export class UpdateBlockContentOperation implements Operation {
  readonly type = OperationType.updateBlock;

  constructor(
    readonly blockIndex: number,
    readonly oldContent: string,
    readonly newContent: string,
  ) {}

  apply(document: Document): Document {
    const block = document.blocks[this.blockIndex];
    return updateBlock(document, this.blockIndex, copyBlock(block, { content: this.newContent }));
  }

  inverse(document: Document): Operation {
    return new UpdateBlockContentOperation(
      this.blockIndex,
      this.newContent,
      this.oldContent,
    );
  }
}

